pandas/core/groupby/ops.py

```diff
@@ -861,15 +868,15 @@ def __iter__(self):
         for i, (start, end) in enumerate(zip(starts, ends)):
             yield i, self._chop(sdata, slice(start, end))
 
-    def _get_sorted_data(self):
+    def _get_sorted_data(self) -> FrameOrSeries:
         return self.data.take(self.sort_idx, axis=self.axis)
 
-    def _chop(self, sdata, slice_obj: slice):
+    def _chop(self, sdata, slice_obj: slice) -> NDFrame:
```

Comment:
why is NDFrame used? is _chop not generic? should DataSplitter be a generic class?

Reply:
I dont understand the question.  Is "generic class" meaningfully different from "base class"?  NDFrame is used because one subclass returns Series and the other returns DataFrame

Reply:
`DataSplitter.__init__` accepts `FrameOrSeries`. do we need to persist this type thoughout the class. i.e. make DataSplitter a generic class. see https://mypy.readthedocs.io/en/latest/generics.html#defining-generic-classes

Reply:
so looking at the definition of _chop in the derived classes, i'm guessing this abstractmethod should be typed as 
```
 def _chop(self, sdata: FrameOrSeries, slice_obj: slice) -> FrameOrSeries:
```

Reply:
Using FrameOrSeries here produces complaints:

```
pandas/core/groupby/ops.py:879: error: Argument 1 of "_chop" is incompatible with supertype "DataSplitter"; supertype defines the argument type as "FrameOrSeries"
pandas/core/groupby/ops.py:879: error: Return type "Series" of "_chop" incompatible with return type "FrameOrSeries" in supertype "DataSplitter"
pandas/core/groupby/ops.py:891: error: Argument 1 of "_chop" is incompatible with supertype "DataSplitter"; supertype defines the argument type as "FrameOrSeries"
pandas/core/groupby/ops.py:891: error: Return type "DataFrame" of "_chop" incompatible with return type "FrameOrSeries" in supertype "DataSplitter"
```

I'm getting close to saying "screw it" when dealing with this type of error.

Reply:
fair enough. probably best to remove the type annotation here then for now. Having one method return FrameOrSeries and another NDFrame seems inconsistent.

Reply:
I think `NDFrame` is correct here; this function isn't generic in the base class rather the subclasses override with the appropriate object

Reply:
mypy won't be looking at the derived classes when it performs type checking. it'll be looking at the type hints on the base class when it checks other methods in the base class.

the abstractmethod should be generic since that is how the derived classes are typed Series -> Series and DataFrame -> DataFrame.

Reply:
I think we are mixing a few different paradigms here. The subclasses should probably be annotated with the type respective to the class, rather than using the TypeVar, i.e. you would never parametrize a `SeriesSplitter` with a DataFrame - it exclusively deals with Series objects

Reply:
> you would never parametrize a `SeriesSplitter` with a DataFrame - it exclusively deals with Series objects

correct. but if a method of the base class is not overridden then the Series type in the derived class will become an NDFrame type after calling that method in the base class.

Reply:
> I think we are mixing a few different paradigms here.

There are some annotations in this PR that make it easier to reason about this code while reading it.  The annotations in this sub-thread are not among them, so I do not particularly care about them.  Let's focus for now on a minimal change needed to get this merged, as there are more bugfix PRs waiting in the wings.

Reply:
https://github.com/pandas-dev/pandas/pull/28339/commits/e6c5f5a31e5a5912bd572e08f03756cbc075752c fixes this.